### ISSUE TITLE
filter_stats does not correctly update index in statsframe

### ISSUE THREAD

**Issue by ilumsden** [LLNL/thicket#147]
`Thicket.filter_stats` will filter the Thicket object based on a predicate applied to the DataFrame in `Thicket.statsframe`. Currently, this function uses `hatchet.GraphFrame.squash` to propagate the filter to other components of the Thicket object (e.g., performance data and graph).

However, because Hatchet has no knowledge of the statsframe, `hatchet.GraphFrame.squash` will not properly update the index of `Thicket.statsframe.dataframe` with `Node` objects from the newly created graph.

As a result, the output of `Thicket.filter_stats` will look alright (i.e., the graph will contain the correct nodes and both the performance and stats dataframes will look correct), but functions that depend on the index of the statsframe (e.g., `Thicket.tree`) will fail with `IndexError`s.

**Comment by ilumsden** [comment]
TLDR

`filter_stats` correctly propagates changes to the graph and performance data. However, it does not update the `Node` objects in the row index of the statsframe's DataFrame. As a result, that DataFrame's row index still uses the old graph's `Node` objects, which are likely to have different `_hatchet_nid` fields than what is in the new graph. This causes functions that rely on the statsframe's row index (e.g., Thicket.tree) to break after calls to `filter_stats`.
